go/new_merkletree/merkle_tree.go

```diff
@@ -0,0 +1,177 @@
+package new_merkletree
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"hash"
+	"math"
+)
+
+// New creates a new merkle hash tree.  The number of "leaves" of the tree,
+// as well as their contents, are retrieved through |dao|.  If you wish to
+// have acceptable performance on non-trivial tree sizes, you'll want to
+// provide |cache| (otherwise, pass `nil`).  The hash function used for all
+// nodes in the tree is specified by |hasher|.
+func New(dao NewMerkleTreeDataInterface, cache NewMerkleTreeCacheInterface, hasher hash.Hash) *NewMerkleTree {
+	return &NewMerkleTree{dao: dao, cache: cache, hasher: hasher}
+}
+
+func (mt NewMerkleTree) CurrentRoot() (Hash, error) {
+	if mt.dao.Size() == 0 {
+		// Special case: empty trees get hashes of empty strings
+		mt.hasher.Reset()
+		return mt.hasher.Sum([]byte{}), nil
+	} else {
+		return mt.subtreeRoot(0, mt.dao.Size()-1)
+	}
+}
+
+func (mt NewMerkleTree) InclusionProof(leaf uint64) ([]Hash, error) {
+	if mt.dao.Size() == 0 {
+		return nil, errors.New("NewMerkleTree: Can't calculate an inclusion proof on an empty tree")
+	}
+
+	if leaf >= mt.dao.Size() {
+		return nil, fmt.Errorf("NewMerkleTree: Invalid leaf entry ID: %v", leaf)
+	}
+
+	return mt.inclusionSubtree(leaf, 0, mt.dao.Size()-1)
+}
+
+func (mt NewMerkleTree) ConsistencyProof(from, to uint64) ([]Hash, error) {
+	if from == 0 {
+		// There's no algorithmic basis for this that I know of, but it is how
+		// existing implementations do it
+		return []Hash{}, nil
+	}
+
+	if to > mt.dao.Size() {
+		return nil, fmt.Errorf("NewMerkleTree.ConsistencyProof: Value for 'to' greater than tree size (to=%v, tree size=%v)", to, mt.dao.Size())
+	}
+
+	if from > to {
+		return nil, fmt.Errorf("NewMerkleTree.ConsistencyProof: 'to' greater than 'from'")
+	}
+
+	return mt.subproof(from, 0, to-1, true)
+}
+
+func (mt NewMerkleTree) hash(s []byte) Hash {
+	mt.hasher.Reset()
+	mt.hasher.Write(s)
+	return mt.hasher.Sum([]byte{})
+}
+
+func (mt NewMerkleTree) leafHash(s []byte) Hash {
+	return mt.hash(bytes.Join([][]byte{{0x0}, s}, []byte{}))
+}
+
+func (mt NewMerkleTree) nodeHash(h1, h2 Hash) Hash {
+	return mt.hash(bytes.Join([][]byte{{0x1}, h1, h2}, []byte{}))
+}
+
+func (mt NewMerkleTree) subtreeRoot(n1, n2 uint64) (Hash, error) {
+	if n1 == n2 {
+		l, err := mt.dao.EntryAt(n1)
+		if err != nil {
+			return nil, err
+		}
+		return mt.leafHash(l), nil
+	} else {
```

Comment:
no need for the `else` block (and elsewhere, I'll stop commenting on them individually)